woocommerce/woocommerce
# Fix: image size customisation controls not shown when theme doesn't define them but Woo core does

**NOTE:** Closed because it diverged too much from master. Superseded by https://github.com/woocommerce/woocommerce/pull/28187.

### All Submissions:


**NOTE:** merging https://github.com/woocommerce/woocommerce/pull/26731 is a prerequisite before this pull request can be merged. Also, base must be changed to `master` before merging.

### Changes proposed in this Pull Request:

In Customize » WooCommerce » Product Images, there should be Main Image Width options, for themes that do not declare image widths. However these controls don't show for the TwentyX themes, since the `includes/theme-support/class-wc-twenty-*.php` classes define them at initialization; the code that renders the controls (in the `WC_Shop_Customizer` class) is simply checking if the settings exist, and can't know who created them.

This pull request introduces the concept of _default_ theme support settings. The theme support classes define the image sizes as defaults, and the code that renders the controls do it only if the image sizes have been defined but not as defaults.

Also, for the cases where the user hasn't explicitly defined image and thumbnail sizes now the default values shown in the editor are the ones defined as theme support (default or not) instead of the hardcoded values of 600 and 300, respectively.

Closes #25900.

### How to test the changes in this Pull Request:

1. Set the Twenty-Twenty theme.
2. Go to Customize » WooCommerce » Product Images and verify that now you can change the image and thumbnail sizes.
3. Add theme support settings for image sizes on the Twenty-Twenty theme and verify that now you can't customize the image and thumbnail sizes. You can add these settings by adding the following code to the `twentytwenty_theme_support` method in `wp-content/themes/twentytwenty/functions.php`: `add_theme_support('woocommerce', array('thumbnail_image_width' => 450,'single_image_width' => 600));`
4. Undo the previous change and delete thumbnail image width size the by running this in the command line: `wp option delete woocommerce_thumbnail_image_width`. Check that the customizer UI now displays for this setting a value of 450, which is the value set in `WC_Twenty_Twenty::init`, instead of the hardcoded value of 300 (if you remove the code that sets the value in `WC_Twenty_Twenty::init` it will fallback to 300).

### Other information:


### Changelog entry

> Controls to change image and thumbnail sizes in theme customisation UI are now displayed even if there's a theme support class defining default values for those.
